storage/innobase/trx/trx0purge.cc

```diff
@@ -428,16 +443,25 @@ trx_purge_truncate_rseg_history(trx_rseg_t &rseg,
   const trx_id_t undo_trx_no=
     mach_read_from_8(b->page.frame + hdr_addr.boffset + TRX_UNDO_TRX_NO);
 
-  if (undo_trx_no >= limit.trx_no)
+  if (undo_trx_no >= trx_no)
   {
-    if (undo_trx_no == limit.trx_no)
-      err = trx_undo_truncate_start(&rseg, hdr_addr.page,
-                                    hdr_addr.boffset, limit.undo_no);
+    if (undo_trx_no == trx_no)
+      err= trx_undo_truncate_start(&rseg, hdr_addr.page,
+                                   hdr_addr.boffset, undo_no);
     goto func_exit;
   }
-
-  if (!all)
-    goto func_exit;
+  else
+  {
+    rseg_ref= rseg.ref_load();
+    if (rseg_ref >= rseg.REF || !purge_sys.sees(rseg.needs_purge))
+    {
+      if (!(rseg.SKIP & rseg_ref) && !freed &&
```

Comment:
I did not want to invoke `rseg.skip_allocation()` because that would lead into two `std::atomic::load()` operations on the `trx_rseg_t::ref`.

The only zero-overhead abstraction that I can think of would be to introduce a helper `class rseg_ref` that wraps `uint32_t`. The `trx_rseg_t::ref` would then be `std::atomic<rseg_ref>` instead of `std::atomic<uint32_t>`. Sure, we could do that, but maybe better in a separate refactoring.

Reply:
I will add a comment that explains the intention. I considered refactoring this code into a separate function, but we would have to pass `rseg_ref` as a parameter to that function.